keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+
+Note: this doc does not discuss `NetworkPolicy`, which cannot currently be used
+to describe a policy that applies to a multi-cluster service.
+
+#### DNS
+
+When a `ServiceExport` is created, this will cause a domain name for the
+multi-cluster service to become accessible from within the supercluster. The
+domain name will be
+`<service-export-name>.<service-export-namespace>.svc.supercluster.local`.
+Requests to this domain name from within the supercluster will resolve to the
+supercluster VIP, which points to the endpoint addresses for pods within the
+underlying `Service`(s) across the supercluster. All service consumers must use
+the `*.svc.supercluster.local` name to enable supercluster routing, even if
+there is a matching `Service` with the same namespaced name in the local
+cluster. There will be no change to existing behavior of the `svc.cluster.local`
+zone.
+
+#### EndpointSlice
+
+When a `ServiceExport` is created, this will cause `EndpointSlice` objects for
+the underlying `Service` to be created in each cluster within the supercluster.
+One or more `EndpointSlice` resources will exist for each cluster that exported
+the `Service`, with each `EndpointSlice` containing only endpoints from its
+source cluster. These `EndpointSlice` objects will be marked as managed by the
+supercluster service controller, so that the endpoint slice controller doesn’t
+delete them.
+
+```
+<<[UNRESOLVED]>>
+We have not yet sorted out scalability impact here. We hope the upper bound for
+imported endpoints + in-cluster endpoints will be ~= the upper bound for
+in-cluster endpoints today, but this remains to be determined.
+<<[/UNRESOLVED]>>
+```
+
+#### Endpoint TTL
+
+To prevent stale endpoints from persisting in the event that a cluster becomes
+unreachable to the supercluster controller, each `EndpointSlice` is associated
+with a lease representing connectivity with its source cluster. The supercluster
+service controller is responsible for periodically renewing the lease so long as
+the connection with the source cluster is confirmed alive. A separate
+controller, that may run inside each cluster, is responsible for watching each
+lease and removing all remaining `EndpointSlices` associated with a cluster when
+that cluster’s lease expires.
+
+#### Service Types
+
+- `ClusterIP`: This is the the straightforward case most of the proposal
+  assumes. Each `EndpointSlice` associated with the exported service is combined
+  with slices from other clusters to make up the supercluster service. They will
+  be imported to the cluster behind the supercluster IP.
+
+```
+<<[UNRESOLVED re:stateful sets]>>
+  Today's headless services likely don't want a VIP and may not function
+  properly behind one. It probably doesn't make sense to export a current
+  headless service to the supercluster, it would work, but likely not the way
+  you want.
+<<[/UNRESOLVED]>>
+```
+- `NodePort` and `LoadBalancer`: These create `ClusterIP` services that would
```

Comment:
Do we have to support all of the types?
This is slightly confusing because I think from the description below, it sounds like you won't be supporting ServiceExport for the NodePort, LoadBalancer types.

Reply:
Reworded and added clarification. I don't see why we wouldn't support these types since they each create ClusterIP services. Another way to put this would be that we only support ClusterIP services, but include the implicit ClusterIP services created by NodePort and LoadBalancer services